pkg/log/logger.go

```diff
@@ -0,0 +1,257 @@
+	for _, sandboxId := range v.workerRunner.ListWorkers() {
+		if _, err := os.Stat(filepath.Join(v.config.KubernetesFolder, sandboxId)); os.IsNotExist(err) {
+			v.workerRunner.StopWorker(sandboxId)
+			continue
+		}
+		if _, err := os.Stat(filepath.Join(v.config.VirtletFolder, sandboxId)); os.IsNotExist(err) {
+			v.workerRunner.StopWorker(sandboxId)
+		}
+	}
+}
+
+const STOP string = "---LOGGER-STOP---"
+
+type workerRunner struct {
+	workers    map[string]chan *tail.Line // map[sandboxId]chan
+	workersMux sync.Mutex
+	statusCh   chan string
+}
+
+func newWorkerRunner() *workerRunner {
+	return &workerRunner{
+		workers:  make(map[string]chan *tail.Line),
+		statusCh: nil,
+	}
+}
+
+func (w *workerRunner) RunNewWorker(inputFile, outputFile, sandboxId string) {
+	fmt.Println("Spawned worker for:", sandboxId)
+
+	w.reportWorkerState("START", sandboxId)
+	defer w.reportWorkerState("STOP", sandboxId)
+
+	if _, err := os.Stat(outputFile); os.IsNotExist(err) {
+		file, err := os.Create(outputFile)
+		if err != nil {
+			fmt.Println("failed to create output file:", err)
+			return
+		}
+		file.Close()
+	}
+
+	f, err := os.OpenFile(outputFile, os.O_WRONLY, 0777)
+	if err != nil {
+		fmt.Println("failed to open output file:", err)
+		return
+	}
+	defer f.Close()
+
+	// Tail VM's file.
+	t, err := tail.TailFile(inputFile, tail.Config{Follow: true, ReOpen: true})
+	if err != nil {
+		fmt.Println("failed to tail input file:", err)
+		return
+	}
+
+	// Expose channel to virtletLogger so that it can close it when needed.
+	w.registerWorker(sandboxId, t.Lines)
+
+	// Do work. This forloop will block until canceled; it will wait for new
+	// lines to come and parse them immediately.
+	for line := range t.Lines {
+		if line.Text == STOP {
+			break
+		}
+
+		if line.Err != nil {
+			fmt.Println("error reading line:", line.Err.Error())
+			break
+		}
+
+		// Convert raw line into Kubernetes json.
+		converted := fmt.Sprintf(`{"time": "%s", "stream": "stdout","log":"%s\n"}`, line.Time.Format(time.RFC3339), escapeLine(line.Text))
+		converted = converted + "\n"
+
+		f.WriteString(converted)
+		f.Sync()
+
+		w.reportWorkerState("RUN", sandboxId)
+	}
+
+	// This code is only executed when t.Lines channel is closed.
+	delete(w.workers, sandboxId)
+	fmt.Printf("Worker for sandbox '%s' stopped gracefully\n", sandboxId)
+}
+
+func (w *workerRunner) ListWorkers() (sandboxIds []string) {
+	for sandboxId, _ := range w.workers {
+		sandboxIds = append(sandboxIds, sandboxId)
+	}
+	return sandboxIds
+}
+
+func (w *workerRunner) WorkerExists(sandboxId string) bool {
+	w.workersMux.Lock()
+	defer w.workersMux.Unlock()
+	return w.workers[sandboxId] != nil
+}
+
+func (w *workerRunner) StopWorker(sandboxId string) {
+	if w.workers[sandboxId] != nil {
+		fmt.Printf("Stop worker '%s'\n", sandboxId)
+		w.workersMux.Lock()
+		defer w.workersMux.Unlock()
+		w.workers[sandboxId] <- &tail.Line{Text: STOP}
+	}
+}
+
+func (w *workerRunner) StopAllWorkers() {
```

Comment:
Am I missing something but this seems to be called only from the tests? Is it actually needed?